### DISCUSSION TITLE
When to use lifespan vs @lru_cache to load data once

### DISCUSSION THREAD

**Discussion by rmtuckerphx** [fastapi/fastapi#11987] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# *****************************************************************
# Example 1: lru_cache
# data is loaded once on first call to get_service
# *****************************************************************
from fastapi import FastAPI, Depends

app = FastAPI()

@lru_cache
def get_service():
  # load data or ML model and pass to service
  # function IS cached
  data = load_large_data()
  return MyService(data)

@app.get("/nlu")
async def get_nlu(text: str, my_service = Depends(get_service):
  # do stuff
  return {"key": "value"}


# *****************************************************************
# Example 2: lifespan
# data is loaded once on startup or just before first request
# *****************************************************************
from fastapi import FastAPI, Depends
from contextlib import asynccontextmanager

my_cache = {}

@asynccontextmanager
async def lifespan(app: FastAPI):  
  print('Lifespan - starting up')
  # load data or ML model and store in cache
  data = load_large_data()
  my_cache['data_key'] = data
  yield
  print('Lifespan - shutting down')
  my_cache.clear()

app = FastAPI(lifespan=lifespan)

def get_service():
  # get data from cache and pass to service
  # function is NOT cached
  data = my_cache['data_key']
  return MyService(data)

@app.get("/nlu")
async def get_nlu(text: str, my_service = Depends(get_service):
  # do stuff
  return {"key": "value"}
```

### Description

When should one use **lifespan** vs **@lru_cache** to load large data once?
What differences (nuances) are there in the two approaches that would lead the choice of one over the other?
Is each function decorated with **@lru_cache** stored in a separate cache?

Both approaches are shown in the code and are marked Example 1 and Example 2.

In the tutorial on **lifespan** it talked of the use case of loading an ML model once on startup which is the case shown in the code of this discussion: large file or ML model are similar enough. I was initially excited about this option because it meant that the model could load on startup and that there would be no delay on the first or any request to the endpoint. But after implementing the approach it seemed that the call to load the data occurred on the first request.

From the perspective of the client using the API, how is **lifespan** loading data on first request any different than **@lru_cache** caching the factory method on first call?

References:
- lru_cache - https://fastapi.tiangolo.com/advanced/settings/?h=#lru_cache-technical-details
- lifespan - https://fastapi.tiangolo.com/advanced/events/#lifespan

### Operating System

Linux

### Operating System Details

Debian GNU/Linux 12 (bookworm)

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

3.11.9

**Comment by ceb10n** [comment, marked as answer]
> From the perspective of the client using the API, how is lifespan loading data on first request any different than @lru_cache caching the factory method on first call?

## lifespan

lifespan can be used when you need to initialize something before your app is ready. Something like: "To serve my services correctly, I need to initialize a connection pool in my db".

Note that with lifespan, it's not on your fist request that this will be initialized, its on startup:

```python
from contextlib import asynccontextmanager

from fastapi import FastAPI


@asynccontextmanager
async def lifespan(app: FastAPI):
  print('Lifespan - starting up')
  yield
  print('Lifespan - shutting down')


app = FastAPI(lifespan=lifespan)


@app.get("/{name}")
def hello(name: str):
  return {"hello": name}
```

You can see it in your logs:

![lifespan](https://github.com/user-attachments/assets/457a5c1f-cfcd-40e0-9a5c-4c323ac18ec8)

## depends with lru

Now if you want to use depends with lru_cache, your dependency will be initialized when the first request comes, not on app startup:

```python
from functools import lru_cache

from fastapi import FastAPI, Depends

app = FastAPI()

@lru_cache
def get_service():
  print("Dependency - starting up")
  return "my cached deps"


@app.get("/{name}")
def hello(name: str, my_dependency = Depends(get_service)):
  print(f"Wow - got my deps: {my_dependency}")
  return {"hello": name}
```

And you can see in the logs that when you app starts, it won't initilize anything. But when the *first request* comes in, it will be initialized. And since your are using `lru_cache`, it will run only once:

![lru cache with depends](https://github.com/user-attachments/assets/d53f0777-1c20-4b61-bfdf-f33f8812502d)


## Depends alone

And if you use depends without `@lru_cache`, it will run *once per request*:

```python
from fastapi import FastAPI, Depends

app = FastAPI()

def get_service():
  print("Dependency - starting up")
  return "my cached deps"


@app.get("/{name}")
def hello(name: str, my_dependency = Depends(get_service)):
  print(f"Wow - got my deps: {my_dependency}")
  return {"hello": name}
```

And you can check in logs that once per request, your deps will be initialized:

![depends without cache](https://github.com/user-attachments/assets/41cc3adf-6c53-4714-a2d3-ae2a5727e2ba)

## Which one is better?

The answer is: it *depends* (LOL)

As my understanding, you can use:

lifespan - You need to initialize something before you app is ready, so when your request comes, it won't be forced to do things that could be done before

depends - Depends works like a charm for request specific things, like your query strings, headers/tokens, etc.

depends with lru - Something that is not crucial for you app to work properly, so you don't want to slow down you app intialization. It can be a singleton, and you don't really care if the first request that needs its dependency will be slower than others.

## answering your question

> But after implementing the approach it seemed that the call to load the data occurred on the first request.

Looking at your code:

```python
my_cache = {}

@asynccontextmanager
async def lifespan(app: FastAPI):  
  print('Lifespan - starting up')
  # load data or ML model and store in cache
  data = load_large_data()
  my_cache['data_key'] = data
  yield
  print('Lifespan - shutting down')
  my_cache.clear()

app = FastAPI(lifespan=lifespan)

def get_service():
  # get data from cache and pass to service
  # function is NOT cached
  data = my_cache['data_key']
  return MyService(data)
```

What is happening is that your are loading your data on app startup, but you are creating a `MyService` instance once per request. I don't know what your are doing in this "service", but that's whats happening: data dictionary loaded in app startup and `MyService` instance is beeing created once per request with the already loaded data.

Let's take a look what is happening with a fake service implemented:

```python
# a global dict
my_cache = {}

# will be called once when your app is starting
def load_large_data():
  print("loading a really large data structure")
  return {"k1": "val1", "k2": "val2"}


class MyService:
  
  def __init__(self, data: dict) -> None:
    self.data = data
    print("And a new instance rises from a request")


# what is before yield will be executed before app is ready
# what is after yield will be executed before app "dies"
@asynccontextmanager
async def lifespan(app: FastAPI):
  print('Lifespan - starting up')
  # load data or ML model and store in cache
  data = load_large_data()
  my_cache['data_key'] = data
  yield
  print('Lifespan - shutting down')
  my_cache.clear()


app = FastAPI(lifespan=lifespan)


# will be executed once per request
def get_service():
  print("Hello, I'm your dependency. I won't load your data, but I'll create a MyService instance with it ;)")
  data = my_cache['data_key']
  return MyService(data)


@app.get("/nlu")
async def get_nlu(my_service = Depends(get_service)):
  print(f"Got my dependency! :D : {my_service}")
  return {"key": "value"}
```

![example](https://github.com/user-attachments/assets/60377ccb-5b3b-4466-8ed2-c96de6128fe8)

**Reply by rmtuckerphx** [reply]
Thanks @ceb10n for the detailed response.

I reimplemented `lifespan` exactly as shown but still no loading on startup.

![image](https://github.com/user-attachments/assets/4020204f-e25d-44b7-bfd3-d043db1f7d13)

I wonder if it has to do with how I'm running the app.

Here is the `docker-compose.yml` file:

```
services:
  app:
    build: .
    container_name: fast-api-services
    command: uvicorn app.src.main:app --host 0.0.0.0 --port 80 --reload
    ports:
      - 80:80
    volumes:
      - .:/code
```

And I'm running it with: `docker compose up --build`
